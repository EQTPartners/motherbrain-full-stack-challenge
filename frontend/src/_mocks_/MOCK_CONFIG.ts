export const MOCK_CONFIG = {
  api: {
    url: 'http://localhost:8080',
  },
  mapbox: {
    apiToken:
      'pk.eyJ1IjoiYWxleHBhdG93IiwiYSI6ImNpZnExMHhhNTZwOHRzNmtxdTVkMnk0azcifQ.lwyfSEFN3WF2rNiRqsLL9A',
  },
};
