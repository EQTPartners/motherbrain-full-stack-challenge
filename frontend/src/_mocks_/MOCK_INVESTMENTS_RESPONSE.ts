export const MOCK_INVESTMENTS_RESPONSE = {
  status: 'ok',
  investor: {
    uuid: '30b3efdf-6024-804d-ea55-c4d1a8dfc86c',
    companyName: 'Goldman Sachs',
    homepageUrl: 'http://www.goldmansachs.com/',
    countryCode: 'USA',
    city: 'New York',
    shortDescription:
      'Goldman Sachs is a multinational financial services firm providing securities, and investment banking and management services.',
    description:
      'The Goldman Sachs Group, Inc. provides investment banking, securities, and investment management services, as well as financial services to corporations, financial institutions, governments, and high-net-worth individuals worldwide. Its Investment Banking segment offers financial advisory services, including advisory assignments with respect to mergers and acquisitions, divestitures, corporate defense, risk management, and restructurings and spin-offs; and underwriting services comprising public offerings and private placements of a range of securities, loans and other financial instruments, and derivative transactions. The company’s Institutional Client Services segment provides client execution services, such as fixed income, currency, and commodities client execution related to making markets in interest rate products, credit products, mortgages, currencies, and commodities; and equities related to making markets in equity products, as well as executes and clears institutional client transactions on stock, options, and futures exchanges. This segment also engages in securities services business providing financing, securities lending, and other brokerage services to institutional clients, including hedge funds, mutual funds, pension funds, and foundations. Its Investing and Lending segment originates longer-term loans; and invests in debt securities, loans, public and private equity securities, real estate, consolidated investment entities, distressed assets, currencies, commodities, and power generation facilities. The company’s Investment Management segment provides investment products and services, as well as offers wealth advisory services, including portfolio management and financial counseling, and brokerage and other transaction services. The Goldman Sachs Group, Inc. was founded in 1869 and is headquartered in New York, New York.',
    fundingRounds: '0',
    fundingTotalUSD: '',
    employeeCount: '10000+',
    location: {
      lat: 40.7148396,
      lng: -74.0144098,
    },
  },
  investments: [
    {
      uuid: '61daf8a8-2d27-2db0-bce7-cac8cc83589e',
      companyName: 'Nav',
      homepageUrl: 'https://www.nav.com/',
      countryCode: 'USA',
      city: 'Draper',
      shortDescription:
        'Nav bridges the gap between SMB and financial institutions by bringing transparency, certainty, and efficiency to B2B credit and financing.',
      description:
        "Nav is the leading Business Financial Management app offering free access to personal and business credit reports from major consumer and commercial credit bureaus including Dun & Bradstreet, Experian and TransUnion. The Nav platform also helps small business owners make informed business credit decisions and hosts a robust marketplace with more than 110 business financing products including loans and credit cards. The marketplace uses a unique, lender-neutral approach to help business owners find the best financing options for their needs. Nav's solution is also leveraged by other business service providers to enhance their customer experience. Nav has offices in Salt Lake City, UT, and Silicon Valley.",
      fundingRounds: '6',
      fundingTotalUSD: '99024727',
      employeeCount: '101-250',
      location: {
        lat: 40.5023071,
        lng: -111.895405,
      },
    },
  ],
};
