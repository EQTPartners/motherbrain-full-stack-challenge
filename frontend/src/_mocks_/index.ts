export * from './MOCK_CONFIG';
export * from './MOCK_INVESTMENTS_RESPONSE';
export * from './MOCK_INVESTORS_RESPONSE';
export * from './MOCK_LINES_DATA';
export * from './MOCK_POINTS_DATA';
