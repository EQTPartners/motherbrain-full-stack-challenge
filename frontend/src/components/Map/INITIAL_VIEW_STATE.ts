export const INITIAL_VIEW_STATE = {
  longitude: -122.2,
  latitude: 37.6,
  zoom: 9.5,
  pitch: 30,
  bearing: 0,
  maxZoom: 13,
  minZoom: 3,
};
