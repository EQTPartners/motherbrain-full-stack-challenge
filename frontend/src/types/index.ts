export * from './Config';
export * from './InvestmentsResponse';
export * from './InvestorsResponse';
export * from './Line';
export * from './LayerTypes';
export * from './Point';
