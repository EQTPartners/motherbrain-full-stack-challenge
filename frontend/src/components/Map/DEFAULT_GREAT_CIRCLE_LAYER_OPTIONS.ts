import { Line } from '../../types';

export const DEFAULT_GREAT_CIRCLE_LAYER_OPTIONS = {
  id: 'great-circle-layer',
  pickable: true,
  getWidth: 12,
  getSourcePosition: (d: Line) => d.from.coordinates,
  getTargetPosition: (d: Line) => d.to.coordinates,
  getSourceColor: (d: Line) => [50, 140, 0, 10],
  getTargetColor: (d: Line) => [50, 140, 0],
  wrapLongitude: true,
};
