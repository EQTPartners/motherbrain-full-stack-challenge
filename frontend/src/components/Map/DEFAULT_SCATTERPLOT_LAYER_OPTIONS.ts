import { Point } from '../../types';

export const DEFAULT_SCATTERPLOT_LAYER_OPTIONS = {
  id: 'scatterplot-layer',
  pickable: true,
  opacity: 0.8,
  stroked: true,
  filled: true,
  radiusScale: 7,
  radiusMinPixels: 1,
  radiusMaxPixels: 200,
  lineWidthMinPixels: 1,
  getPosition: (d: Point) => d.coordinates,
  getRadius: (d: Point) => 150,
  getFillColor: (d: Point) => [255, 140, 0],
  getLineColor: (d: Point) => [0, 0, 0, 0],
};
