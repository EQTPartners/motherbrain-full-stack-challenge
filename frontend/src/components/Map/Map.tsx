import React, { useState } from 'react';
// @ts-ignore
import { GreatCircleLayer, ScatterplotLayer } from 'deck.gl';
// @ts-ignore
import DeckGL from '@deck.gl/react';
import { StaticMap } from 'react-map-gl';

import { DEFAULT_GREAT_CIRCLE_LAYER_OPTIONS } from './DEFAULT_GREAT_CIRCLE_LAYER_OPTIONS';
import { DEFAULT_SCATTERPLOT_LAYER_OPTIONS } from './DEFAULT_SCATTERPLOT_LAYER_OPTIONS';
import { INITIAL_VIEW_STATE } from './INITIAL_VIEW_STATE';
import { getConfig } from '../../conifg';
import { InvestmentsResponse, InvestorsResponse, Line, MapTypes, Point } from '../../types';
import { MapDataFormatter } from '../../utils';

interface MapProps {
  data: InvestorsResponse | InvestmentsResponse | null;
  handleInvestorClick: (name: string) => void;
  type: MapTypes;
}

const isLine = (object: any): object is Line => object.to != null;

export const Map = (props: MapProps) => {
  const { data, handleInvestorClick, type } = props;
  const {
    mapbox: { apiToken },
  } = getConfig();
  const [tooltipProperties, setTooltipProperties] = useState<{
    object: Point | Line | null;
    x: number;
    y: number;
  }>({ object: null, x: 0, y: 0 });

  const getLayer = () => {
    if (data == null) {
      return;
    }

    if (type === MapTypes.Investors) {
      return getInvestorsLayer();
    }

    return getInvestmentsLayer();
  };

  const getInvestorsLayer = () => {
    const formattedData = MapDataFormatter.convertInvestorsResponseToPointData(
      data as InvestorsResponse,
    );

    return new ScatterplotLayer({
      ...DEFAULT_SCATTERPLOT_LAYER_OPTIONS,
      data: formattedData,
      onHover: ({ object, x, y }: { object: Point; x: number; y: number }) => {
        setTooltipProperties({ object, x, y });
      },
      onClick: ({ object }: { object: Point }) => handleInvestorClick(object.name),
    });
  };

  const getInvestmentsLayer = () => {
    const formattedData = MapDataFormatter.convertInvestmentsResponseToLineData(
      data as InvestmentsResponse,
    );

    return new GreatCircleLayer({
      ...DEFAULT_GREAT_CIRCLE_LAYER_OPTIONS,
      data: formattedData,
      onHover: ({ object, x, y }: { object: Line; x: number; y: number }) => {
        setTooltipProperties({ object, x, y });
      },
    });
  };

  const renderTooltip = () => {
    const { object, x, y } = tooltipProperties;

    if (object == null) {
      return null;
    }

    const message = isLine(object) ? object.to.name : object.name;

    return (
      <div
        style={{
          position: 'absolute',
          zIndex: 1,
          pointerEvents: 'none',
          left: x + 10,
          top: y + 10,
        }}
      >
        {message}
      </div>
    );
  };

  return (
    <DeckGL
      initialViewState={INITIAL_VIEW_STATE}
      controller={true}
      layers={[getLayer()]}
      width={'100%'}
      height={'100%'}
    >
      <StaticMap mapboxApiAccessToken={apiToken} width={'100%'} height={'100%'} />
      {renderTooltip()}
    </DeckGL>
  );
};
