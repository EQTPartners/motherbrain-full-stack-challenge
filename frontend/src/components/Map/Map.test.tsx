import React from 'react';
import { render } from '@testing-library/react';

import { Map } from './Map';
import { MOCK_INVESTMENTS_RESPONSE } from '../../_mocks_';
import { MapTypes } from '../../types';

describe('Map', () => {
  // Rendering the Deck.GL components isn't working because GL only exists in the browser.
  it.skip('should render', () => {
    const { container } = render(
      <Map
        data={MOCK_INVESTMENTS_RESPONSE}
        handleInvestorClick={() => {}}
        type={MapTypes.Investments}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
