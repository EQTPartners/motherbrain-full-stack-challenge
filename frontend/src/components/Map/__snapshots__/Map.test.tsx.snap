// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map should render 1`] = `
<div
  id="deckgl-wrapper"
  style="position: absolute; z-index: 0; left: 0px; top: 0px; width: 100%; height: 100%;"
>
  <canvas
    id="deckgl-overlay"
    style="left: 0px; top: 0px; width: 100%; position: absolute; height: 100%;"
  />
</div>
`;
